Automattic/woocommerce-payments
# Always save subscriptions payment methods when using UPE

Fixes #3471

#### Changes proposed in this Pull Request

WCPay offers a setting to disable "payments via saved cards." But for subscriptions, we always want to save the payment method regardless of that setting, in order to process recurring payments. This is currently the case for non-UPE checkout, but not for UPE checkout. This PR updates the UPE payment gateway to always save the payment method if it's processing a recurring payment.

#### Testing Instructions

The following steps will verify that, when saved cards is disabled, a payment method is saved for subscription products with UPE, but not for non-subscription products.

1. Enable WCPay Subscriptions if not already enabled.
2. In **Payments > Settings**, ensure **Enabled payments via saved cards** is unchecked, saving changes if needed.
3. In **WCPay Dev**, ensure **Enable UPE checkout** is checked, submitting changes if needed.
4. Create a new customer account and sign in. In that customer account, purchase a subscription product with a test card number.
5. As the customer, go to `/my-account/subscriptions/` and click on the subscription you just purchased. Verify that, next to **Payment**, you see details of your payment method (e.g. "Visa ending in 4242"). Before this PR's changes, you can verify that you instead saw something generic like "Credit card / debit card".
6. Check the database's `wp_woocommerce_payment_tokens` table and verify that there's a payment token associated with this customer's user ID.
7. Delete the saved payment method (as by enabling saved cards, cancelling the subscription, deleting the payment method at `/my-account/payment-methods/`, then disabling saved cards again). Verify that the token no longer appears in the `wp_woocommerce_payment_tokens` table.
8. Purchase a non-subscription product with a test card number.
9. Check the database's `wp_woocommerce_payment_tokens` table and verify that there was no new payment token added for the customer.
10. If you re-enable saved cards and go to `/my-account/payment-methods/` again, there should be no payment method listed.

-------------------


**Post merge**
